Minimal Suse Support, developed and tested on SLES11 SP3

buzzdeee opened the pull request:
Very basic support to be able to use kmod::load, loads and unloads
modules correctly, and updates the /etc/sysconfig/kernel file
accordingly. Other defined types etc. weren't touched since
not needed for now by me.

Add some specs for module_load for SLES, but they are not picked
up. With SLES added to metadata.json, travis showed messages like:
Can't find facts for 'sles-11.3-x86_64' for facter 1.7, skipping...

raphink commented on manifests/load.pp:
```diff
@@ -18,7 +18,16 @@
     'present': {
-      $changes = "clear '${name}'"
+      case $::osfamily {
+        'Debian': {
```
IIRC, this also covers RedHat.

raphink commented on manifests/load.pp:
```diff
@@ -27,7 +36,16 @@
-      $changes = "rm '${name}'"
+      case $::osfamily {
+        'Debian': {
+          $changes = "rm '${name}'"
```
Same here.

mcanevet commented on manifests/load.pp:
```diff
@@ -18,7 +18,16 @@
+          $changes = "set MODULES_LOADED_ON_BOOT/value[last()+1] '${name}'"
+          $onlyif  = "match MODULES_LOADED_ON_BOOT/value[.='${name}'] include ${name}"
+        }
+        default: { }
```
You should probably fail here

buzzdeee replied:
failing would fail in the RedHat case, that I deliberately did not added here.

mcanevet commented on manifests/load.pp:
```diff
@@ -27,7 +36,16 @@
+          $changes = "rm MODULES_LOADED_ON_BOOT/value[.='${name}']"
+          $onlyif  = "match MODULES_LOADED_ON_BOOT/value[.='${name}'] not_include ${name}"
+        }
+        default: { }
```
Same here

buzzdeee replied:
failing would fail in the RedHat case, that I deliberately did not added here.

mcanevet replied:
@buzzdeee why did you deliberately not added here?

buzzdeee replied:
Am 7/13/2015 um 11:16 AM schrieb Mickaël Canévet:
>
> In manifests/load.pp 
> <https://github.com/camptocamp/puppet-kmod/pull/35#discussion_r34445468>:
>
> > @@ -27,7 +36,16 @@
> >      }
> >
> >      'absent': {
> > -      $changes = "rm '${name}'"
> > +      case $::osfamily {
> > +        'Debian': {
> > +          $changes = "rm '${name}'"
> > +        }
> > +        'Suse': {
> > +          $changes = "rm MODULES_LOADED_ON_BOOT/value[.='${name}']"
> > +          $onlyif  = "match MODULES_LOADED_ON_BOOT/value[.='${name}'] not_include ${name}"
> > +        }
> > +        default: { }
>
> @buzzdeee <https://github.com/buzzdeee> why did you deliberately not 
> added here?
>

The $changes is not used by RedHat, therefore, I kept the default empty 
here, instead of failing.
Otherwise, I could add an empty RedHat statement, and fail in the 
default case.

Sebastian

> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/camptocamp/puppet-kmod/pull/35/files#r34445468>.
>

mcanevet commented:
@buzzdeee you should also add Suse support in metadata.json so that unit tests are run on it.

buzzdeee commented:
Added SLES to metadata.json, however, the spec tests don't get picked up, as described in the initial pull request message, see the travis logs, don't know what's going wrong there, since I added what facter reports:
facter | grep operating
operatingsystem => SLES
operatingsystemrelease => 11.3

mcanevet commented:
@buzzdeee use `11` not `11.3` in `operatingsystemrelease`

buzzdeee commented:
@mcanevet thanks, using 11 instead of 11.3 worked

buzzdeee commented:
rebased and squashed onto your latest master

raphink commented:
@buzzdeee have you seen my inline comments?

buzzdeee commented:
I saw the inline comments, and I was sure, I replied. Maybe I only intended to do so, and forgot, sorry.

What I see in the load.pp, is that the $changes, before I added Suse support, was only used by Debian, since Debian uses Augeas lense, like I implemented for Suse, but RedHat using a template, and there is no $changes in the template as far as I could see. 

Because of that, I think RedHat doesn't need the $changes at all.

raphink commented:
@buzzdeee I see, makes sense

raphink commented on manifests/load.pp:
```diff
@@ -18,7 +18,16 @@
+        }
+        'Suse': {
+          $changes = "set MODULES_LOADED_ON_BOOT/value[last()+1] '${name}'"
+          $onlyif  = "match MODULES_LOADED_ON_BOOT/value[.='${name}'] not_include ${name}"
```
You could make this idempotent by doing this:

```puppet
"set MODULES_LOADED_ON_BOOT/value[.='${name}'] '${name}'"
```

without an `onlyif` statement.

raphink commented on manifests/load.pp:
```diff
@@ -27,7 +36,16 @@
+        }
+        'Suse': {
+          $changes = "rm MODULES_LOADED_ON_BOOT/value[.='${name}']"
+          $onlyif  = "match MODULES_LOADED_ON_BOOT/value[.='${name}'] include ${name}"
```
`rm` doesn't need `onlyif`. It's already idempotent this way.

raphink commented:
Here, a few more comments on the Augeas changes.

buzzdeee commented:
Rebased to your latest master.
Updated load.pp according to the suggestions, and further updated the spec tests, which are still green.

Haven't tried it yet on a Suse box, will do so on Monday when I'm back at work.
When it works there, I'll squash and report back.

Sebastian

raphink commented:
OK, I'll wait for your tests then.

buzzdeee commented:
As promised, tested on a SLES box today, and seems to be fine.
Squashed again.

mcanevet commented on manifests/load.pp:
```diff
@@ -18,7 +18,15 @@
+        'Suse': {
+          $changes = "set MODULES_LOADED_ON_BOOT/value[.='${name}'] '${name}'"
+        }
+        default: { }
```
What about `RedHat`?

buzzdeee replied:
The $changes is set for Augeas. RedHat uses a template file, and therefore has no need for that $changes variable at all. See the whole file, at the bottom.
The default: case is empty, to please puppet-lint. If wanted, I could add a comment into that default: statement, to make it more clear.

same applies to the similar snipped in the 'absent' case below.

mcanevet commented on manifests/load.pp:
```diff
@@ -27,7 +35,15 @@
+        'Suse': {
+          $changes = "rm MODULES_LOADED_ON_BOOT/value[.='${name}']"
+        }
+        default: { }
```
Same here
